### PR TITLE
Add playwright tests and add to CI

### PR DESCRIPTION
**Pull request by ddimaria**
Addresses issue https://github.com/libp2p/js-libp2p-webrtc/issues/86

- [x] Add Playwright test for the browser-to-server example
- [x] Integrate Playwright tests in CI

### CODE REVIEW

File: examples/browser-to-server/index.js [hunk 1]
```diff
@@ -9,7 +9,7 @@ import { pushable } from 'it-pushable';
 let stream;
 const output = document.getElementById('output')
 const sendSection = document.getElementById('send-section')
-const appendOutput = (line) => output.innerText += `${line}\n`
+const appendOutput = (line) => output.innerHTML += `<div>${line}</div>`
```
Comment (MarcoPolo) [inline examples/browser-to-server/index.js:12]: Why this change? Isn't this unnecessarily a XSS attack vector now? It's an example, yes. But people will copy this.

**Review by MarcoPolo — CHANGES_REQUESTED** [review]
One small thing. Otherwise looks good

File: examples/browser-to-server/tests/test.js [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+import { playwright } from 'test-util-ipfs-example'
+import { spawn, exec } from 'child_process'
+import { existsSync } from 'fs'
+
+// Setup
+const play = test.extend({
+  ...playwright.servers()
+})
+
+async function spinUpGoLibp2p() {
```
Comment (ckousik) [inline examples/browser-to-server/tests/test.js:12]: nit: spawnGoLibp2p?

File: examples/browser-to-server/tests/test.js [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+  if (!existsSync('../../examples/go-libp2p-server/go-libp2p-server')) {
+    await new Promise((resolve, reject) => {
+      exec('go build',
+        { cwd: '../../examples/go-libp2p-server' },
+        (error, stdout, stderr) => {
+          if (error) {
+            reject(error)
+            console.error(`exec error: ${error}`)
+            return
```
Comment (ckousik) [inline examples/browser-to-server/tests/test.js:21]: should this throw an error?

File: examples/browser-to-server/tests/test.js [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+  const server = spawn('./go-libp2p-server', [], { cwd: '../../examples/go-libp2p-server', killSignal: 'SIGINT' })
+  server.stderr.on('data', (data) => {
+    console.log(`stderr: ${data}`, typeof data)
+  })
+  const serverAddr = await (new Promise(resolve => {
+    server.stdout.on('data', (data) => {
+      console.log(`stdout: ${data}`, typeof data)
+      if (data.includes('p2p addr:  ')) {
+        // Parse the addr out
+        resolve((String(data)).match(/p2p addr:  ([^\s]*)/)[1])
```
Comment (ckousik) [inline examples/browser-to-server/tests/test.js:37]:
```
        const addr = String(data).match(/p2p addr: ([^\s]*)/)
        if (addr.length === 0) {
          <throw error>
        }
        resolve(addr[1])
```

File: examples/browser-to-server/index.js (hunk 1 above)
Comment (ddimaria) [reply examples/browser-to-server/index.js]: Good call, this has refactored to be less permissive.

File: examples/browser-to-server/tests/test.js (hunk 2 above)
Comment (ddimaria) [reply examples/browser-to-server/tests/test.js:12]: Good name change, this has been updated.

File: examples/browser-to-server/tests/test.js (hunk 3 above)
Comment (ddimaria) [reply examples/browser-to-server/tests/test.js:21]: I just checked the various approached and throwing the error results in better error output.  Thanks @ckousik!

File: examples/browser-to-server/tests/test.js (hunk 4 above)
Comment (ddimaria) [reply examples/browser-to-server/tests/test.js:37]: I had to change the suggestion as we get (and expect) output that doesn't include the multiaddr that isn't an error case.  Additionally, we need to check for the null case when there are no matches.

```
const addr = String(data).match(/p2p addr:  ([^\s]*)/)
if (addr !== null && addr.length > 0) {
  resolve(addr[1])
}   
```

File: examples/browser-to-server/package.json [hunk 5]
```diff
@@ -5,14 +5,20 @@
   "type": "module",
   "scripts": {
     "start": "vite",
-    "go-libp2p-server": "cd ../go-libp2p-server && go build && ./go-libp2p-server"
+    "build": "vite build",
+    "go-libp2p-server": "cd ../go-libp2p-server && go build && ./go-libp2p-server",
```
Comment (GlenDC) [inline examples/browser-to-server/package.json:9]: Why actually going through the trouble of building it and running it, instead of just using `go run ./main.go` or w/e the binary entry is?

File: examples/browser-to-server/tests/test.js [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+// Setup
+const play = test.extend({
+  ...playwright.servers()
+})
+
+async function spawnGoLibp2p() {
+  if (!existsSync('../../examples/go-libp2p-server/go-libp2p-server')) {
+    await new Promise((resolve, reject) => {
+      exec('go build',
```
Comment (GlenDC) [inline examples/browser-to-server/tests/test.js:15]: same here, might as well just always run it, no? instead of building it first?

File: examples/browser-to-server/package.json [hunk 7]
```diff
@@ -5,14 +5,20 @@
...
+    "test": "npm run build && playwright test tests"
   },
   "dependencies": {
     "@chainsafe/libp2p-noise": "^11.0.0",
     "@multiformats/multiaddr": "^11.0.5",
-    "js-libp2p-webrtc": "../../",
     "it-pushable": "^3.1.0",
+    "js-libp2p-webrtc": "../../",
```
Comment (GlenDC) [inline examples/browser-to-server/package.json:16]: actual package is now called @libp2p/webrtc, so we might as well call it here this as well.

File: examples/browser-to-server/package.json [hunk 8]
```diff
@@ -5,14 +5,20 @@
   "type": "module",
```
Comment (GlenDC) [inline examples/browser-to-server/package.json:5]: I think it falls out of scope for this PR, but something to think about is that we might as well turn this repo into a workspace. This way the example and webrtc package could be both living in the same package, allowing the example to make use of the @libp2p/webrtc package the same as any external usage would. Making the example also directly copyable without having to rely on relative paths in its package.json to make it work given that the relative paths would be fixed by the root workspace definition :) Just my 2 cents on this matters, but feels cleaner to me.

File: examples/browser-to-server/tests/test.js [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+        resolve(addr[1])
+      }        
+    })
+  }))
+  return { server, serverAddr }
+}
+
+play.describe('bundle ipfs with parceljs:', () => {
+  // DOM
+  const connectBtn = '#connect'
```
Comment (GlenDC) [inline examples/browser-to-server/tests/test.js:44]: 2 comments:

- is there no convention for constants, in many language it's something like `UPPER_CASE` but dunno about ts or PL in general. Either way I would expect it to be different from regular variable names?
- might want to use a convention suffix for these ones to make clear that these are Html Element Identifier strings rather than actual "buttons", "addresses", "input fields", etc...?

File: examples/browser-to-server/tests/test.js [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+  const output = '#output'
+
+  let server
+  let serverAddr
+
+  // eslint-disable-next-line no-empty-pattern
+  play.beforeAll(async ({ }, testInfo) => {
+    testInfo.setTimeout(5 * 60_000)
+    const s = await spawnGoLibp2p()
+    server = s.server
```
Comment (GlenDC) [inline examples/browser-to-server/tests/test.js:57]: dunno what the convention is but unpacking `{ server, serverAddress} = s` should work here, no?

File: examples/browser-to-server/tests/test.js [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+    serverAddr = s.serverAddr
+    console.log('Server addr:', serverAddr)
+  }, {})
+
+  play.afterAll(() => {
+    server.kill('SIGINT')
+  })
+
+  play.beforeEach(async ({ servers, page }) => {
+    console.log(`http://localhost:${servers[0].port}/`)
```
Comment (GlenDC) [inline examples/browser-to-server/tests/test.js:67]: When I log and use such a composed value I usually assign it to a const first, but no biggie

File: examples/browser-to-server/tests/test.js [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+    
+    // add the go libp2p multiaddress to the input field and submit
+    await page.fill(connectAddr, serverAddr)
+    await page.click(connectBtn)
+
+    // send the relay message to the go libp2p server
+    await page.fill(messageInput, message)
+    await page.click(sendBtn)
+
+    await page.waitForSelector('#output:has(div)')
```
Comment (GlenDC) [inline examples/browser-to-server/tests/test.js:82]: again no biggie but I notice a mix of single and double quotes in this file for no good reason, but given the linter doesn't complain I suppose there is no strict def to be used here?

File: examples/browser-to-server/tests/test.js (hunk 9 above)
Comment (MarcoPolo) [reply examples/browser-to-server/tests/test.js:44]: >  convention suffix for these ones to make clear that these are Html Element Identifier strings rather than actual "buttons"

That's what the types are for :)

File: examples/browser-to-server/tests/test.js (hunk 6 above)
Comment (MarcoPolo) [reply examples/browser-to-server/tests/test.js:15]: Good question! this should be documented. `go run ./main.go` builds the program then starts a subprocess for the program. SIGINT on the parent process doesn't stop the subprocess (try running `go run` in the background and issuing a `kill -2 pid`). By running the binary directly we avoid the indirection and SIGINT works as expected.

I wonder if we could SIGKILL this instead and it would work?

I don't think it's worth changing in this PR.

File: examples/browser-to-server/tests/test.js (hunk 10 above)
Comment (MarcoPolo) [reply examples/browser-to-server/tests/test.js:57]: No. This is assigning to closured variables.

File: examples/browser-to-server/package.json (hunk 7 above)
Comment (MarcoPolo) [inline examples/browser-to-server/package.json:16]:
```suggestion
    "@libp2p/webrtc": "../../",
```

File: examples/browser-to-server/tests/test.js (hunk 6 above)
Comment (GlenDC) [reply examples/browser-to-server/tests/test.js:15]: I agree that it is probably out of scope. More of a note.

File: examples/browser-to-server/tests/test.js (hunk 9 above)
Comment (GlenDC) [reply examples/browser-to-server/tests/test.js:44]: What types? O.O

Comment (GlenDC) [reply examples/browser-to-server/tests/test.js:44]: You mean that this could be implied because it is of type str? Dunno, not entirely sure that is the case. Certainly not for the output one imho. I like to be very clear about this thing, but neither do I want to make a foothold here, so I'm fine with whatever you end up choosing here, just  note from my POV.

File: examples/browser-to-server/tests/test.js (hunk 10 above)
Comment (GlenDC) [reply examples/browser-to-server/tests/test.js:57]: Hmmmm, ok.

File: examples/browser-to-server/tests/test.js (hunk 9 above)
Comment (MarcoPolo) [reply examples/browser-to-server/tests/test.js:44]: Oh my mistake, I thought this was a `.ts` file

File: examples/browser-to-server/package.json (hunk 5 above)
Comment (ddimaria) [reply examples/browser-to-server/package.json:9]: Thanks Glen, I added this change.

File: examples/browser-to-server/package.json (hunk 8 above)
Comment (ddimaria) [reply examples/browser-to-server/package.json:5]: The pathing has been removed in favor of the npm package import.  The example was written before the package was released, so this was something that needed to be done now that it's published.  Thanks for the reminder!  

Regarding the placement of this comment, is there a concern with `"type": "module"`?

File: examples/browser-to-server/tests/test.js (hunk 9 above)
Comment (ddimaria) [reply examples/browser-to-server/tests/test.js:44]: This was a borrow from the WebTransport Playwright test, though Hungarian(ish) notation may be beneficial here.  I'm OK leaving as is.

File: examples/browser-to-server/package.json (hunk 8 above)
Comment (GlenDC) [reply examples/browser-to-server/package.json:5]: No no, just decided to put it on top of the file it was related to. Sorry for the confusion there.

File: examples/browser-to-server/tests/test.js (hunk 10 above)
Comment (ddimaria) [reply examples/browser-to-server/tests/test.js:57]: Yes, these are singletons where the variables are scoped in the `play.describe` closure.  The styleguide for this repo doesn't use semicolons, so doing the recommended destructuging assignment on an already defined variable causes issues:
```
({ server, serverAddr } = await spawnGoLibp2p())
```
The parser appends this to the `setTimeout()` function call above it (even with an extra line break), resulting in a TypeError: `TypeError: testInfo.setTimeout(...) is not a function`.

Adding semicolons resolves the issue but makes the code incompatible with the main `js-libp2p` styles and one that Aegir expects in lints.

I'm going to leave as is.

Comment (GlenDC) [reply examples/browser-to-server/tests/test.js:57]: More then fair enough. Sorry for the trouble 😰

File: examples/browser-to-server/tests/test.js (hunk 11 above)
Comment (ddimaria) [reply examples/browser-to-server/tests/test.js:67]: I like this suggestion ❤️   It's implemented.

File: examples/browser-to-server/tests/test.js (hunk 12 above)
Comment (ddimaria) [reply examples/browser-to-server/tests/test.js:82]: `"hello"` is now `'hello'` on line 73.

**Review by MarcoPolo — APPROVED** [review]
